modules/switches.py

```diff
@@ -0,0 +1,107 @@
+import random
+import modules
+import enum
+
+class Switches(modules.Module):
+    class D():
+        UP=False
+        DOWN=True
+    identifiers = ['switches']
+    display_name = "Switches"
+    manual_name = "Switches"
+    help_text = "`{cmd} flip 1 5 3 4 2 5 2` or `{cmd} flip 1534252`. Partial solutions allowed."
+    module_score = 4
+
+    # each number is a switch in the up state
+    invalid_positions = ["3","245","2345","14","145","1345","12","124","123","1234"]
+    positions_for_svg_generation = [[50.5, 174], [65.5, 174], [73, 234], [43, 234]]
+
+    # lists indexes of switches in the up position
+    def up_state_positions(self, switch_list):
+        return "".join([str(i+1) for i, val in enumerate(switch_list) if val == self.D.UP])
+
+    # checks if the state is valid
+    def is_valid_state(self, switch_state):
+        return not self.up_state_positions(switch_state) in self.invalid_positions
+
+    def gen_rand(self):
+        return [random.choice([self.D.DOWN, self.D.UP]) for i in range(5)]
+
+    def __init__(self, bomb, ident):
+        super().__init__(bomb, ident)
+
+        self.solution = self.gen_rand()
+        while not self.is_valid_state(self.solution):
+            self.solution = self.gen_rand()
+
+        self.switches = self.gen_rand()
+        while True:
+            if self.is_valid_state(self.switches) and self.switches != self.solution:
+                break
+            self.switches = self.gen_rand()
+
+        init_str=" ".join(map(lambda x:("UP","DOWN")[x],self.switches))
```

Comment:
I feel like this should be refactored into a separate function: `state_as_string`. You could then use it to add some logging after each switch being flipped.

Reply:
yeah that might be helpful